src/python/pants/backend/javascript/package_json.py

```diff
@@ -0,0 +1,633 @@
+    pass
+
+
+@rule
+async def all_first_party_node_package_targets(targets: AllTargets) -> FirstPartyNodePackageTargets:
+    return FirstPartyNodePackageTargets(
+        tgt for tgt in targets if tgt.has_fields((PackageJsonSourceField, NodePackageNameField))
+    )
+
+
+@dataclass(frozen=True)
+class OwningNodePackageRequest:
+    address: Address
+
+
+@dataclass(frozen=True)
+class OwningNodePackage:
+    target: Target | None = None
+    third_party: tuple[Target, ...] = ()
+
+    @classmethod
+    def no_owner(cls) -> OwningNodePackage:
+        return cls()
+
+    def ensure_owner(self) -> Target:
+        if self != OwningNodePackage.no_owner():
+            assert self.target
+            return self.target
+        raise ValueError("No owner could be determined.")
+
+
+@rule
+async def find_owning_package(request: OwningNodePackageRequest) -> OwningNodePackage:
+    candidate_targets = await Get(
+        Targets,
+        RawSpecs(
+            ancestor_globs=(AncestorGlobSpec(request.address.spec_path),),
+            description_of_origin=f"the `{OwningNodePackage.__name__}` rule",
+        ),
+    )
+    package_json_tgts = sorted(
+        (tgt for tgt in candidate_targets if tgt.has_field(PackageJsonSourceField)),
+        key=lambda tgt: tgt.address.spec_path,
+        reverse=True,
+    )
+    tgt = package_json_tgts[0] if package_json_tgts else None
+    if tgt:
+        deps = await Get(Targets, DependenciesRequest(tgt[Dependencies]))
+        return OwningNodePackage(
+            tgt, tuple(dep for dep in deps if dep.has_field(NodeThirdPartyPackageNameField))
+        )
+    return OwningNodePackage()
+
+
+@rule
+async def parse_package_json(content: FileContent) -> PackageJson:
+    parsed_package_json = FrozenDict.deep_freeze(json.loads(content.content))
+    return PackageJson(
+        content=parsed_package_json,
+        name=parsed_package_json["name"],
+        version=parsed_package_json["version"],
+        snapshot=await Get(Snapshot, PathGlobs([content.path])),
+        module=parsed_package_json.get("type"),
+        workspaces=tuple(parsed_package_json.get("workspaces", ())),
+        dependencies=FrozenDict.deep_freeze(
+            {
+                **parsed_package_json.get("dependencies", {}),
+                **parsed_package_json.get("devDependencies", {}),
+                **parsed_package_json.get("peerDependencies", {}),
+            }
+        ),
+    )
+
+
+@rule
+async def read_package_jsons(globs: PathGlobs) -> PackageJsonForGlobs:
+    snapshot = await Get(Snapshot, PathGlobs, globs)
+    digest_contents = await Get(DigestContents, Digest, snapshot.digest)
+    return PackageJsonForGlobs(
+        await MultiGet(
+            Get(PackageJson, FileContent, digest_content) for digest_content in digest_contents
+        )
+    )
+
+
+@rule
+async def all_package_json() -> AllPackageJson:
+    return AllPackageJson(await Get(PackageJsonForGlobs, PathGlobs(["**/package.json"])))
```

Comment:
So, while attempting to run `fmt` on this branch after adjusting the timeouts, I encountered two issues here:

1. This line will match all `package.json` files in a repository, regardless of whether they have had `package_json` targets have been tailored for them. That means that it finds the `package.json` file that we have in `docs` (which, for now, we're fine having ignored). Instead, what you probably want to do is request `AllTargets`, and then filter them to the appropriate type: e.g. https://github.com/pantsbuild/pants/blob/213b19fa23f9281f4f43a63f03ac02f3c4825ff1/src/python/pants/jvm/dependency_inference/artifact_mapper.py#L255-L259
2. `docs/package.json` does not have a `name` or `version` field, which causes `parse_package_json` to fail with a keyerror, since it considers them to be required. AFAICT from https://docs.npmjs.com/cli/v9/configuring-npm/package-json, `name` and `version` are only required if you intend to publish a package.

Just fixing 1 is likely sufficient, but it would probably be good to improve the error for 2.

Reply:
>  That means that it finds the package.json file that we have in docs (which, for now, we're fine having ignored).

Oh, hm. It looks like this is not committed to the repository, which explains why you / CI did not see it.

Reply:
Thanks... not a blocker for landing, so I'm leaving auto-merge on.